File: backend/app/schemas/f2f.py
# app/schemas/f2f.py

"""
Authoritative Face-to-Face (F2F) request/response contract.

Consolidated 2026-08-22: this module previously held a stale, unused,
minimal F2FCreateRequest (never imported by any router, test, or
Alembic script — confirmed by repo-wide search). The complete, active
contract actually used by app/api/f2f.py was instead defined inline in
that file. This module now holds the single authoritative definition;
app/api/f2f.py imports from here. No field, type, default, or
validator was changed — this is a pure move, not a behavior change.
"""

from __future__ import annotations

from datetime import date
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


class F2FCreateRequest(BaseModel):
    patient_id: UUID
    benefit_period_id: UUID
    encounter_date: date

    # performed_by_role is NOT accepted from the request — the endpoint
    # always derives it from the authenticated user's own role, so a
    # caller can never record an unauthorized discipline as the F2F
    # performer. Retained here only for backward-compatible clients;
    # any client-supplied value is ignored.
    performed_by_role: Optional[str] = None

    summary: Optional[str] = Field(default=None, max_length=5000)
    clinical_decline_summary: Optional[str] = Field(default=None, max_length=5000)

    # Functional scoring
    kps_score: Optional[int] = None
    pps_score_previous: Optional[int] = None
    pps_score_current: Optional[int] = None
    ecog_score_previous: Optional[int] = Field(default=None, ge=0, le=5)
    ecog_score_current: Optional[int] = Field(default=None, ge=0, le=5)

    # Disease scoring
    fast_score: Optional[str] = None
    nyha_class: Optional[str] = None

    # ADL / decline
    adl_dependency_level: Optional[str] = None
    adl_dependency_count: Optional[int] = None
    is_bedbound: Optional[bool] = None

    # Objective decline indicators
    weight_loss_lbs: Optional[float] = None
    oral_intake_decline: Optional[bool] = None
    dysphagia: Optional[bool] = None
    hospitalizations_30d: Optional[int] = None
    oxygen_lpm_previous: Optional[float] = None
    oxygen_lpm_current: Optional[float] = None

    primary_diagnosis: Optional[str] = None
    secondary_conditions: Optional[str] = None


class F2FCreateResponse(BaseModel):
    id: UUID
    status: str
    encounter_date: date


class F2FFinalizeRequest(BaseModel):
    # Used when NP performed the F2F and physician review/attestation is captured on the encounter.
    attestation_summary: Optional[str] = Field(default=None, max_length=5000)


class F2FFinalizeResponse(BaseModel):
    id: UUID
    status: str
    finalized_at: Optional[str]

File: sns-emr-frontend/src/api/f2f.ts
import api from "./client";

export type F2FStatus = "DRAFT" | "FINALIZED";

export type F2FRecord = {
  id: string;
  patient_id: string;
  benefit_period_id: string;
  status: F2FStatus;
  status_label: string;
  encounter_date: string | null;
  performed_by_role: string | null;
  performed_by_name: string | null;
  attesting_provider_name: string | null;
  attested_at: string | null;
  summary: string | null;
  finalized_at: string | null;
};

export type F2FCreateRequest = {
  patient_id: string;
  benefit_period_id: string;
  encounter_date: string;
  summary?: string;
  clinical_decline_summary?: string;
  kps_score?: number;
  pps_score_previous?: number;
  pps_score_current?: number;
  fast_score?: string;
  nyha_class?: string;
  adl_dependency_level?: string;
  adl_dependency_count?: number;
  is_bedbound?: boolean;
  weight_loss_lbs?: number;
  oral_intake_decline?: boolean;
  dysphagia?: boolean;
  hospitalizations_30d?: number;
  oxygen_lpm_previous?: number;
  oxygen_lpm_current?: number;
  primary_diagnosis?: string;
  secondary_conditions?: string;
};

export async function listF2FEncounters(patientId: string): Promise<F2FRecord[]> {
  const response = await api.get<F2FRecord[]>(`/f2f/patients/${patientId}`);
  return response.data || [];
}

export async function createF2FEncounter(payload: F2FCreateRequest) {
  const response = await api.post(`/f2f/`, payload);
  return response.data;
}

export async function finalizeF2FEncounter(f2fId: string, attestationSummary?: string) {
  const response = await api.post(`/f2f/${f2fId}/finalize`, { attestation_summary: attestationSummary });
  return response.data;
}

export async function getF2FStatusHistory(f2fId: string) {
  const response = await api.get(`/f2f/${f2fId}/status-history`);
  return response.data || [];
}
